### ISSUE TITLE
`attn_mask` in nn.MultiheadAttention is additive

### ISSUE THREAD

**Issue by mttk** [pytorch/pytorch#21518]
## 📚 Documentation

It likely should be mentioned that the `attn_mask` argument of MHA is an additive mask (-inf masks values), rather than the standard multiplicative mask (0 masks values). Perhaps even enforce a value check (all values should be 0 / -inf ?, otherwise print warning?)

**Comment by zou3519** [comment]
cc @zhangguanheng66

**Comment by zhangguanheng66** [comment]
We have an open PR now to integrate torch.nn.MultiheadAttention moduel with the one in Fairseq. Will fix the docs together. Thanks for the suggestion.

**Comment by mttk** [comment]
Thanks, didn't see that PR when searching. This can be closed as far as I'm concerned.

**Comment by zhangguanheng66** [comment]
> Thanks, didn't see that PR when searching. This can be closed as far as I'm concerned.

I have linked the PR. Will close this issue automatically once we land the PR.

**Comment by bhattg** [comment]
Now will the MultiheadedAttention throw error if any element of attn_mask is not 0 or -inf ?

**Comment by pmixer** [comment]
Hi @zhangguanheng66 , for tracking the additive attn_mask versus multiplicative attn_mask issue, I tested by the following code:

```python

attn_mask = torch.tril(torch.ones((8,8)))
inputs = torch.ones((8, 2, 6))
mha = torch.nn.MultiheadAttention(6, 2) # hidden_dim=6, head_num=2
outputs, weights = mha(inputs, inputs, inputs, attn_mask=attn_mask) # Q, K, V, attn_mask for causality
```
to enforce causality, but the returned attention weights suggest it still attend to future inputs:

```
 tensor([[[0.2797, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029],
          [0.2377, 0.2377, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874],
          [0.2066, 0.2066, 0.2066, 0.0760, 0.0760, 0.0760, 0.0760, 0.0760],
          [0.1828, 0.1828, 0.1828, 0.1828, 0.0672, 0.0672, 0.0672, 0.0672],
          [0.1638, 0.1638, 0.1638, 0.1638, 0.1638, 0.0603, 0.0603, 0.0603],
          [0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.0546, 0.0546],
          [0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.0499],
          [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]],
 
         [[0.2797, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029],
          [0.2377, 0.2377, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874],
          [0.2066, 0.2066, 0.2066, 0.0760, 0.0760, 0.0760, 0.0760, 0.0760],
          [0.1828, 0.1828, 0.1828, 0.1828, 0.0672, 0.0672, 0.0672, 0.0672],
          [0.1638, 0.1638, 0.1638, 0.1638, 0.1638, 0.0603, 0.0603, 0.0603],
          [0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.0546, 0.0546],
          [0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.0499],
          [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]]],
        grad_fn=<DivBackward0>))
```

does that means its still additive mask in current implementation(I used PyTorch 1.6.0+cu101 on google colab)?

THX!

**Comment by zhangguanheng66** [comment]
> Hi @zhangguanheng66 , for tracking the additive attn_mask versus multiplicative attn_mask issue, I tested by the following code:
> 
> ```python
> attn_mask = torch.tril(torch.ones((8,8)))
> inputs = torch.ones((8, 2, 6))
> mha = torch.nn.MultiheadAttention(6, 2) # hidden_dim=6, head_num=2
> outputs, weights = mha(inputs, inputs, inputs, attn_mask=attn_mask) # Q, K, V, attn_mask for causality
> ```
> 
> to enforce causality, but the returned attention weights suggest it still attend to future inputs:
> 
> ```
>  tensor([[[0.2797, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029],
>           [0.2377, 0.2377, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874],
>           [0.2066, 0.2066, 0.2066, 0.0760, 0.0760, 0.0760, 0.0760, 0.0760],
>           [0.1828, 0.1828, 0.1828, 0.1828, 0.0672, 0.0672, 0.0672, 0.0672],
>           [0.1638, 0.1638, 0.1638, 0.1638, 0.1638, 0.0603, 0.0603, 0.0603],
>           [0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.0546, 0.0546],
>           [0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.0499],
>           [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]],
>  
>          [[0.2797, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029, 0.1029],
>           [0.2377, 0.2377, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874, 0.0874],
>           [0.2066, 0.2066, 0.2066, 0.0760, 0.0760, 0.0760, 0.0760, 0.0760],
>           [0.1828, 0.1828, 0.1828, 0.1828, 0.0672, 0.0672, 0.0672, 0.0672],
>           [0.1638, 0.1638, 0.1638, 0.1638, 0.1638, 0.0603, 0.0603, 0.0603],
>           [0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.1485, 0.0546, 0.0546],
>           [0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.1357, 0.0499],
>           [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]]],
>         grad_fn=<DivBackward0>))
> ```
> 
> does that means its still additive mask in current implementation(I used PyTorch 1.6.0+cu101 on google colab)?
> 
> THX!

I think your `attn_mask` is not set up correctly. For the LM task, you can take a look at [generate_square_subsequent_mask](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html?highlight=generate_square_subsequent_mask#torch.nn.Transformer.generate_square_subsequent_mask). `attn_mask` in MHA supports three types and a float mask will be added to the attention weight. You might want to try a bool mask. The doc are [here](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention.forward). Let me know if you still have questions.

**Comment by pmixer** [comment]
> ```python
> attn_mask = torch.tril(torch.ones((8,8)))
> inputs = torch.ones((8, 2, 6))
> mha = torch.nn.MultiheadAttention(6, 2) # hidden_dim=6, head_num=2
> outputs, weights = mha(inputs, inputs, inputs, attn_mask=attn_mask) # Q, K, V, attn_mask for causality
> ```

THX! I tried converting attn_mask to `torch.bool` type and used `mha(inputs, inputs, inputs, attn_mask=~attn_mask) # Q, K, V, attn_mask for causality` for applying the mask, then it works(in torch 1.6, did not work in torch 1.4, seems a recent fix). Thanks for your help and have a good day :)

**Comment by TheEngineerProgrammer** [comment]
Actually attn_mask support 2 types of tensor.
According to de doc: Binary and float masks are supported. For a binary mask, a True value indicates that the corresponding position is not allowed to attend. For a float mask, the mask values will be added to the attention weight. 

So in the first case you can use something like this:
attn_mask = torch.tril(torch.ones((8,8))) == 0
inputs = torch.ones((8, 2, 6))
mha = torch.nn.MultiheadAttention(6, 2) # hidden_dim=6, head_num=2
mha(inputs, inputs, inputs, attn_mask=attn_mask)

Here the mask is like this (but bigger):
False True True True
False False True True
False False False True
False False False False

And in the second case you can use something like this:
attn_mask = torch.triu(torch.ones((8,8)),diagonal=1)*(-1e9)
inputs = torch.ones((8, 2, 6))
mha = torch.nn.MultiheadAttention(6, 2) # hidden_dim=6, head_num=2
mha(inputs, inputs, inputs, attn_mask=attn_mask)

Here the mask is like this (but bigger):
0 -1e9 -1e9 -1e9
0 0 -1e9 -1e9
0 0 0 -1e9
0 0 0 0
